### DISCUSSION TITLE
Does setting branch protection rules need too broad permissions?

### DISCUSSION THREAD

**Discussion by yngvark** [community/community#59060] [Apps, API and Webhooks]
### Select Topic Area

Product Feedback

### Body

I'm in the process of creating a GitHub app, that needs permissions to set a repository's branch protection rules.

As far as I can see, according to https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-administration

the app needs "repository administration" permission in order to do that.

Unfortunately, this permission gives the app access to create new repositories. I believe this is overly broad permissions. I want to be able to change settings on specific, existing repositories, not create new repositories.

I could not find any discussion about this out there, so I'm just posting it to learn if my thinking is right, and as a feedback to the GitHub team.

(As an implementation detail, I'm using the [github Terraform provider](https://registry.terraform.io/providers/integrations/github/latest/docs/resources/branch_protection.html) when implementing this.)

**Comment by patrick-knight** [comment]
That is accurate info. The ability to manage branch protections can be done via [custom roles](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-peoples-access-to-your-organization-with-roles/about-custom-repository-roles#repository), which are not designed for apps however.

I'd love to learn more about how you're look to leverage an app for managing branch protections?

**Reply by yngvark** [reply]
> I'd love to learn more about how you're look to leverage an app for managing branch protections?

The Terraform resource I'm using is this: https://registry.terraform.io/providers/integrations/github/latest/docs/resources/branch_protection_v3

I'm configuring the GitHub provider to authenticate as an app like this:

```tf
# https://registry.terraform.io/providers/integrations/github/latest/docs#github-app-installation
provider "github" {
  owner = "myorg"

  app_auth {
    id              = var.app_id
    installation_id = var.app_installation_id
    pem_file        = var.app_pem_file
  }
}
```

That's all. When running `terraform apply`, the provider authenticates as an app, and thus commits made to a repository come from the app (instead of my personal user). I guess that should answer your question?

**Reply by patrick-knight** [reply]
Thanks @yngvark that's helpful feedback!
